### ISSUE TITLE
Oddworld: Stranger's Wrath [NPEB00645] Graphical glitches, vertex issues

### ISSUE THREAD

**Issue by LoStraniero91** [RPCS3/rpcs3#6270]
**RPCS3 version:** 0.0.6-8420-230c3d55 Alpha
**System specs:** i9-9900k/32 GB/RTX 2080 Ti/Windows 10.0.17763

**Note:** Tested with game version 1.00 (the original release), the issues still apply with 1.02 and probably 2.00 (which is a pre-patched rerelease with fixes and additional features).

The game suffers from graphical glitches in form of overblooming (in neon signs, lights, water and other effects) and stretched vertex in all characters (all seen in the screenshot).
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3437938/RPCS3.log.gz)

![Immagine](https://user-images.githubusercontent.com/29802222/61991375-7696a580-b04f-11e9-9da8-8ed923a1049c.png)

**Comment by benderscruffy** [comment]
still happens on both open gl and vulkan
![oddworld](https://user-images.githubusercontent.com/2696404/66309212-9eea2c80-e93b-11e9-848b-1ce2f370e406.PNG)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3697034/RPCS3.log.gz)

**Comment by LoStraniero91** [comment]
Still valid as 0.0.7-9195-46ca39ec
![STR](https://user-images.githubusercontent.com/29802222/70068792-ef56d100-15f0-11ea-876a-13ad655ba427.png)

**Comment by LoStraniero91** [comment]
Still valid as 0.0.8-9487-7d2ed920 
![Immagine](https://user-images.githubusercontent.com/29802222/73600583-a40a6180-4552-11ea-9bb1-912ae2251633.png)

Provided a RSX Capture, even if by loading it back with RPCS3, I get even weirder graphical glitches. 
![Immagined](https://user-images.githubusercontent.com/29802222/73600663-e3857d80-4553-11ea-9977-ae93be2830d0.png)
https://mega.nz/#!4mR3XIaK!ctd4A02h8oCxyLW8398OW0XxksRxdlzU5U6UaugqWE4

**Comment by benderscruffy** [comment]
still happens in the latest master
here is a couple of renderdocs
https://mega.nz/file/gyRjGQZB#zDh1gp3Kjh5eu0zDObnMSKJSZFvj4V5yTODfoZ_qLOs

**Comment by Asinin3** [comment]
Still an issue?

**Comment by LoStraniero91** [comment]
0.0.18-12862

Yes, still an issue.

![image](https://user-images.githubusercontent.com/29802222/135766422-8d90db75-f81f-4e4f-991f-f7bbecf93a89.png)

**Comment by LoStraniero91** [comment]
Still an issue in **0.0.25-14373**

Do you need some updated RSX captures?

**Comment by kd-11** [comment]
Wouldn't hurt to add. Though usually vertex issues are usually caused by SPU problems which is why this is not on my queue.

**Comment by LoStraniero91** [comment]
Status is unchanged in RPCS3 0.0.37-18022.

<img width="1919" height="1032" alt="Image" src="https://github.com/user-attachments/assets/c97c258a-96de-479b-bc87-50c607906ec4" />

**Comment by ghost** [comment]
Can we get an update on this?

**Comment by LoStraniero91** [comment]
> Can we get an update on this?

Unchanged as of **0.0.38-18238**

Like kd-11 said, it's likely a SPU issue.

<img width="1919" height="1030" alt="Image" src="https://github.com/user-attachments/assets/f043ce98-10bf-4bfd-809f-a940fe38242a" />

**Comment by Jack54guythecoder** [comment]
Hey, is there an update to this?

**Comment by Jack54guythecoder** [comment]
@LoStraniero91 Did you rechecked this issue again?

**Comment by ChwaraeTeg** [comment]
Issue still present in 0.0.40-19260

**Comment by LoStraniero91** [comment]
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/b779d7eb-53b5-48d1-98bd-db93153d934e" />

Status is unchanged in 0.0.42-19675. This is one hell of a holy grail bug.

**Comment by Jack54guythecoder** [comment]
Yeah, tho apparently it is a SPU issue, so maybe more work needs to be done on the SPU component....?

**Comment by Jack54guythecoder** [comment]
The suspicious code is in SPU float conversions and bit-manipulation routines, especially the recompiler paths?

**Comment by Jack54guythecoder** [comment]
Recommended investigation order
FOUND OUT BY AI, btw.
Trace the game under the SPU interpreter and record whether FI executes immediately before the bad geometry is generated.
Compare FI results between:
SPU interpreter
LLVM recompiler
ASMJIT recompiler
If FI matches, repeat for:
MPY / MPYS
FMA/FMS/FNMA/FNMS
FRSQEST
Add temporary checks for:
NaNs
extended SPU values
infinities
unusually large finite values
position vectors whose magnitude exceeds a practical threshold
Use the existing RSX capture to identify whether the bad position is already present in the vertex buffer or is introduced by the RSX vertex shader.

**Comment by Jack54guythecoder** [comment]
ONE LAST THING, MAYBE THIS COULD BE USEFUL AS WELL, I TRY TO BE.....SOMETIMES.....

Thanks for keeping this reported.

Based on the symptom pattern here — stretched vertices / exploded character geometry that persists across builds — this looks much more like an SPU floating-point correctness bug than a normal RSX shader issue. The most likely candidates are the SPU float interpolation / multiply / FMA paths, especially:

FI (floating interpolate)
MPY / MPYS / FMA-family ops
possibly FREST / FRSQEST if normalization is involved
I’d suggest the following next steps to narrow it down:

**Compare the same frame under:**

SPU interpreter
LLVM recompiler
ASMJIT recompiler If the bad vertex data appears only under one backend, that strongly narrows the bug.
Instrument the SPU float ops around vertex transform / interpolation / animation math and log:

offending input/output registers
NaN / denormal / extended-float cases
any overflow or exponent-mismatch conditions
Check whether the bad values already exist before RSX vertex processing, or only after the SPU-generated transform data reaches RSX.

Add one more RSX capture and a fresh log with:

Accurate XFloat enabled
SPU verification enabled
the exact build/version used
Specifically look for cases where a position or weight is turning into an extreme finite value or NaN-like bit pattern, because that would explain the stretched/exploded mesh effect.

If this is truly SPU-side, this should be fixable by targeting the erroneous float handling in the recompiler/interpreter rather than changing graphics settings.

This would also explain why the issue has been persistent across multiple versions and why it does not look like a simple driver/rendering bug.
